From 7ab9f41a87febda9fea136980b7c5b8d5e1eaaad Mon Sep 17 00:00:00 2001
From: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Date: Thu, 27 Sep 2018 10:23:33 -0700
Subject: [PATCH 0430/1852] drm: xlnx: pl_disp: Pass interlaced field info to
 frmbuf DMA

commit 476fe74e2177ad45a77e8211d38cfdcfebbce138 from
https://github.com/Xilinx/linux-xlnx.git

This patch passes interlaced field information to frmbuf DMA
based on framebuffer flag.

Signed-off-by: Satish Kumar Nagireddy <satish.nagireddy.nagireddy@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_pl_disp.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/xlnx_pl_disp.c b/drivers/gpu/drm/xlnx/xlnx_pl_disp.c
index d8c8d7e00467..28f5779ac6a2 100644
--- a/drivers/gpu/drm/xlnx/xlnx_pl_disp.c
+++ b/drivers/gpu/drm/xlnx/xlnx_pl_disp.c
@@ -63,6 +63,7 @@ struct xlnx_dma_chan {
  * @drm: core drm object
  * @fmt: drm color format
  * @vtc_bridge: vtc_bridge structure
+ * @fid: field id
  */
 struct xlnx_pl_disp {
 	struct device *dev;
@@ -76,6 +77,7 @@ struct xlnx_pl_disp {
 	struct drm_device *drm;
 	u32 fmt;
 	struct xlnx_bridge *vtc_bridge;
+	u32 fid;
 };
 
 /*
@@ -178,6 +180,17 @@ static void xlnx_pl_disp_plane_enable(struct drm_plane *plane)
 	desc->callback = xlnx_pl_disp->callback;
 	desc->callback_param = xlnx_pl_disp->callback_param;
 
+	if (plane->state->fb->flags == DRM_MODE_FB_ALTERNATE_TOP ||
+	    plane->state->fb->flags == DRM_MODE_FB_ALTERNATE_BOTTOM) {
+		if (plane->state->fb->flags == DRM_MODE_FB_ALTERNATE_TOP)
+			xlnx_pl_disp->fid = 1;
+		else
+			xlnx_pl_disp->fid = 0;
+
+		xilinx_xdma_set_fid(xlnx_dma_chan->dma_chan, desc,
+				    xlnx_pl_disp->fid);
+	}
+
 	dmaengine_submit(desc);
 	dma_async_issue_pending(xlnx_dma_chan->dma_chan);
 }
-- 
2.31.1

